### PR TITLE
Reduce steady-state allocations in ui_stack_system

### PR DESCRIPTION
**Pull request by UkoeHB**
# Objective

- Reduce allocations on the UI hot path.

## Solution

- Cache buffers used by the `ui_stack_system`.

## Follow-Up

- `sort_by_key` is potentially-allocating. It might be worthwhile to include the child index as part of the sort-key and use unstable sort.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> For some reason my rust analyzer is erroring on the `Label` in `bevy_ui/stack.rs`, with message "`const STORAGE_TYPE` is not a member of trait `Component`".

Do a `cargo clean`: this was recently changed.

**Review by james7132 — COMMENTED** [review]
Generally looks good, though I'd like to see some actual benchmarks. `many_buttons` and the other UI stress tests are probably the best scenarios to check.

I'm not too happy with the idea of a Vec per non-leaf node being allocated here, even in the cached steady state. Ideally we would just use one giant Vec and retain ranges, but this can be attempted in a follow up PR.

File: crates/bevy_ui/src/stack.rs [hunk 1]
```diff
@@ -15,6 +15,24 @@ pub struct UiStack {
     pub uinodes: Vec<Entity>,
 }
 
+#[derive(Default)]
+pub struct StackingContextCache {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/stack.rs:19]: This needs docs if it's public. I would probably add docs for internal users and doc(hidden) this.

**Comment by alice-i-cecile** [comment]
@james7132 I don't think this was actually breaking: the types weren't publicly exported despite the `pub` visibility.

**Review by james7132 — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Pull request merged**